core/src/main/java/org/apache/iceberg/io/CachingFileIO.java

```diff
@@ -0,0 +1,363 @@
+    private final InputFile wrappedInputFile;
+
+    private CachingInputFile(ContentCache cache, InputFile inFile) {
+      this.contentCache = cache;
+      this.wrappedInputFile = inFile;
+    }
+
+    @Override
+    public long getLength() {
+      CacheEntry buf = contentCache.cache.getIfPresent(location());
+      return (buf != null) ? buf.length() : wrappedInputFile.getLength();
+    }
+
+    /**
+     * Opens a new {@link SeekableInputStream} for the underlying data file, either through cache or through the
+     * inner FileIO.
+     * <p>
+     * If data file is not cached yet, and it can fit in the cache, the file content will be cached first before
+     * returning a {@link ByteBufferInputStream}. Otherwise, return a new SeekableInputStream from the inner FIleIO.
+     *
+     * @return a {@link ByteBufferInputStream} if file exist in the cache or can fit in the cache. Otherwise, return a
+     * new SeekableInputStream from the inner FIleIO.
+     */
+    @Override
+    public SeekableInputStream newStream() {
+      try {
+        // read-through cache if file length is less than or equal to maximum length allowed to cache.
+        if (getLengthChecked() <= contentCache.maxContentLength()) {
+          return cachedStream();
+        }
+
+        // fallback to non-caching input stream.
+        return wrappedInputFile.newStream();
+      } catch (FileNotFoundException e) {
+        throw new NotFoundException(e, "Failed to open input stream for file: %s", wrappedInputFile.location());
+      } catch (IOException e) {
+        throw new RuntimeIOException(e, "Failed to open input stream for file: %s", wrappedInputFile.location());
+      }
+    }
+
+    @Override
+    public String location() {
+      return wrappedInputFile.location();
+    }
+
+    @Override
+    public boolean exists() {
+      CacheEntry buf = contentCache.cache.getIfPresent(location());
+      return buf != null || wrappedInputFile.exists();
+    }
+
+    public long getLengthChecked() throws IOException {
+      try {
+        return getLength();
+      } catch (RuntimeIOException ex) {
+        throw ex.getCause();
+      }
+    }
+
+    private CacheEntry getFileContent() {
+      try {
+        int fileLength = (int) getLength();
+        int totalBytesToRead = fileLength;
+        SeekableInputStream stream = wrappedInputFile.newStream();
+        List<ByteBuffer> buffers = Lists.newArrayList();
+
+        while (totalBytesToRead > 0) {
+          // read the stream in 8MB chunk
+          int bytesToRead = Math.min(BUFFER_CHUNK_SIZE, totalBytesToRead);
+          byte[] buf = new byte[bytesToRead];
+          int bytesRead = IOUtil.readRemaining(stream, buf, 0, bytesToRead);
+          totalBytesToRead -= bytesRead;
+
+          if (bytesRead < bytesToRead) {
+            // We are reading less than we should, possibly hitting EOF.
+            // Truncate buf and break from the loop.
+            buf = Arrays.copyOf(buf, bytesRead);
```

Comment:
Rather than copying, can't this just create the buffer with a smaller limit?

Reply:
Do you mean to use second variant of `ByteBuffer.wrap`? Like `ByteBuffer.wrap(buf, 0, bytesRead)`?
I suppose we can, but that does not seem to reduce the backing array size?

Reply:
I'm keeping the copying method in [1b98dc2](https://github.com/apache/iceberg/pull/4518/commits/1b98dc218a0e7b2ab66dead1aa2a40b623aaba3e), but reduce the chunk size from 8MB to 4MB.

The intention is that, if we indeed unexpectedly hit EOF early, the worst case is we are doing slightly under 4MB copy rather than slightly under 8MB copy. But if `getLength` is accurate, we should not reach this copying code.